### PR TITLE
NMS-13346: Add commands to examine telemetryd

### PR DESCRIPTION
**Pull request by fooker**
This adds two commands to examine the configuration and the internal
status of telemetryd. First command listens the configured listeners,
second one lists the parsers defined for a listener. The parsers can
show the internal parser specific state if requested.

### External References

* JIRA (Issue Tracker): http://issues.opennms.org/browse/NMS-13346

### CODE REVIEW

**Comment by fooker** [comment]
Output example:

```
admin@opennms()> opennms:telemetry-listeners 
Name = Multi-UDP-9999
Description = UDP *:9999
Properties:
  Max Packet Size = 8096
  Port = 9999
Parsers:
  - Multi-UDP-9999.Netflow-5-Parser
  - Multi-UDP-9999.Netflow-9-Parser
  - Multi-UDP-9999.IPFIX-TCP-Parser
  - Multi-UDP-9999.SFlow-Parser

admin@opennms()> opennms:telemetry-parsers -s Multi-UDP-9999
Name = Multi-UDP-9999.Netflow-5-Parser
Description = Netflow v5
Properties:
  Clock Skew Event Rate = 3600
  Dns Lookups Enabled = true
  Illegal Flow Event Rate = 3600
  Max Clock Skew = 0
  Sequence Number Patience = 32
  Template Timeout = 1800
  Threads = 16
State:
  Exporters:

Name = Multi-UDP-9999.Netflow-9-Parser
Description = Netflow v9
Properties:
  Clock Skew Event Rate = 3600
  Dns Lookups Enabled = true
  Illegal Flow Event Rate = 3600
  Max Clock Skew = 0
  Sequence Number Patience = 32
  Template Timeout = 1800
  Threads = 16
State:
  Exporters:
    - Key = 10.10.10.10#123456789
      Templates:
        - Template Id = 266
          Age = 22
        - Template Id = 256
          Age = 22
      Options:
        - Template Id = 256
          Age = 22
          Selectors:
          Values:
            - Name = SAMPLING_ALGORITHM
              Value = 2
            - Name = SAMPLING_INTERVAL
              Value = 1

Name = Multi-UDP-9999.IPFIX-TCP-Parser
Description = IPFix
Properties:
  Clock Skew Event Rate = 3600
  Dns Lookups Enabled = true
  Illegal Flow Event Rate = 3600
  Max Clock Skew = 0
  Sequence Number Patience = 32
  Template Timeout = 1800
  Threads = 16
State:
  Exporters:

Name = Multi-UDP-9999.SFlow-Parser
Description = SFlow
Properties:
  Dns Lookups Enabled = true
  Threads = 16
```

File: features/telemetry/protocols/netflow/parser/src/main/java/org/opennms/netmgt/telemetry/protocols/netflow/parser/session/TcpSession.java [hunk 1]
```diff
@@ -98,12 +103,12 @@ public Template lookupTemplate(final int templateId) throws MissingTemplateExcep
     }
 
     private final static class TemplateKey {
-        public final long observationDomainId;
```
Comment (christianpape) [inline features/telemetry/protocols/netflow/parser/src/main/java/org/opennms/netmgt/telemetry/protocols/netflow/parser/session/TcpSession.java:101]: Why was this renamed?

File: features/telemetry/protocols/sflow/parser/src/main/java/org/opennms/netmgt/telemetry/protocols/sflow/parser/SFlowUdpParser.java [hunk 2]
```diff
@@ -179,6 +179,16 @@ public String getName() {
...
     }
 
+    @Override
+    public String getDescription() {
+        return "SFlow";
+    }
+
+    @Override
+    public Object dumpInternalState() {
+        return null;
```
Comment (christianpape) [inline features/telemetry/protocols/sflow/parser/src/main/java/org/opennms/netmgt/telemetry/protocols/sflow/parser/SFlowUdpParser.java:189]: Can you create issues for the unimplemented methods for SFlow, BMP, and so on?

File: features/telemetry/shell/src/main/java/org/opennms/netmgt/telemetry/shell/Format.java [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.stream.Stream;
+import java.util.stream.StreamSupport;
+
+import org.apache.commons.lang3.StringUtils;
+
+import com.google.gson.JsonArray;
+import com.google.gson.JsonObject;
+
+public enum Format {
```
Comment (christianpape) [inline features/telemetry/shell/src/main/java/org/opennms/netmgt/telemetry/shell/Format.java:41]: Really nice class!

**Review by christianpape — APPROVED** [review]
Only a few suggestions.

File: features/telemetry/protocols/netflow/parser/src/main/java/org/opennms/netmgt/telemetry/protocols/netflow/parser/session/TcpSession.java (hunk 1 above)
Comment (fooker) [reply features/telemetry/protocols/netflow/parser/src/main/java/org/opennms/netmgt/telemetry/protocols/netflow/parser/session/TcpSession.java]: Reverted.

File: features/telemetry/protocols/sflow/parser/src/main/java/org/opennms/netmgt/telemetry/protocols/sflow/parser/SFlowUdpParser.java (hunk 2 above)
Comment (fooker) [reply features/telemetry/protocols/sflow/parser/src/main/java/org/opennms/netmgt/telemetry/protocols/sflow/parser/SFlowUdpParser.java:190]: There is no internal state for SFlow but I've added a minimal state for BMP.

Comment (fooker) [reply features/telemetry/protocols/sflow/parser/src/main/java/org/opennms/netmgt/telemetry/protocols/sflow/parser/SFlowUdpParser.java:190]: And added a comment to the SFlow implementation.

**Review by christianpape — APPROVED** [review]
Nice.
